src/bodygraph.cpp

```diff
@@ -0,0 +1,577 @@
+}
+
+const std::vector<bodygraph> &bodygraph::get_all()
+{
+    return bodygraph_factory.get_all();
+}
+
+void bodygraph::finalize_all()
+{
+    bodygraph_factory.finalize();
+}
+
+void bodygraph::check_all()
+{
+    bodygraph_factory.check();
+}
+
+void bodygraph::load( const JsonObject &jo, const std::string & )
+{
+    optional( jo, was_loaded, "parent_bodypart", parent_bp );
+    optional( jo, was_loaded, "fill_sym", fill_sym );
+    if( jo.has_string( "fill_color" ) ) {
+        fill_color = color_from_string( jo.get_string( "fill_color" ) );
+    }
+
+    if( !was_loaded || jo.has_array( "rows" ) ) {
+        rows.clear();
+        for( const JsonValue jval : jo.get_array( "rows" ) ) {
+            if( !jval.test_string() ) {
+                jval.throw_error( "\"rows\" array must contain string values." );
+            } else {
+                std::wstring wrow = utf8_to_wstr( jval.get_string() );
+                std::vector<std::string> row;
+                for( const wchar_t wc : wrow ) {
+                    row.emplace_back( wstr_to_utf8( std::wstring( 1, wc ) ) );
+                }
+                rows.emplace_back( row );
+            }
+        }
+    }
+
+    if( !was_loaded || jo.has_object( "parts" ) ) {
+        parts.clear();
+        for( const JsonMember memb : jo.get_object( "parts" ) ) {
+            const std::string sym = memb.name();
+            JsonObject mobj = memb.get_object();
+            bodygraph_part bpg;
+            optional( mobj, false, "body_parts", bpg.bodyparts );
+            optional( mobj, false, "sub_body_parts", bpg.sub_bodyparts );
+            optional( mobj, false, "sym", bpg.sym, fill_sym );
+            if( mobj.has_string( "select_color" ) ) {
+                bpg.sel_color = color_from_string( mobj.get_string( "select_color" ) );
+            } else {
+                bpg.sel_color = fill_color;
+            }
+            optional( mobj, false, "nested_graph", bpg.nested_graph );
+            parts.emplace( sym, bpg );
+        }
+    }
+}
+
+void bodygraph::finalize()
+{
+    if( rows.size() > BPGRAPH_MAXROWS ) {
+        debugmsg( "body_graph \"%s\" defines more rows than the maximum (%d).", id.c_str(),
+                  BPGRAPH_MAXROWS );
+    }
+
+    int width = -1;
+    bool w_warned = false;
+    for( std::vector<std::string> &r : rows ) {
+        int w = r.size();
+        if( width == -1 ) {
+            width = w;
+        } else if( !w_warned && width != w ) {
+            debugmsg( "body_graph \"%s\" defines rows with different widths.", id.c_str() );
+            w_warned = true;
+        } else if( !w_warned && w > BPGRAPH_MAXCOLS ) {
+            debugmsg( "body_graph \"%s\" defines rows with more columns than the maximum (%d).", id.c_str(),
+                      BPGRAPH_MAXCOLS );
+            w_warned = true;
+        }
+        for( int i = ( BPGRAPH_MAXCOLS - w ) / 2; i > 0; i-- ) {
+            r.insert( r.begin(), " " );
+        }
```

Comment:
```suggestion
        r.insert( r.begin(), ( BPGRAPH_MAXCOLS - w ) / 2, " " );
```
same for the other loop-insert blocks